File: sdks/java/src/main/java/com/cavs/sdk/nativebridge/FfmNativeBridge.java
package com.cavs.sdk.nativebridge;

import java.lang.foreign.Arena;
import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.SymbolLookup;
import java.lang.foreign.ValueLayout;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

/**
 * {@link NativeBridge} backed by the Java Foreign Function &amp; Memory API
 * (JEP 454, finalized in Java 22). It needs no third-party dependency: the
 * whole binding is JDK-native.
 */
public final class FfmNativeBridge implements NativeBridge {

    private static final Linker LINKER = Linker.nativeLinker();

    private final MethodHandle version;
    private final MethodHandle abiVersion;
    private final MethodHandle capabilities;
    private final MethodHandle ctxNew;
    private final MethodHandle ctxFree;
    private final MethodHandle setProgress;
    private final MethodHandle execute;
    private final MethodHandle startJson;
    private final MethodHandle jobPoll;
    private final MethodHandle jobCancel;
    private final MethodHandle jobFree;
    private final MethodHandle resultJson;
    private final MethodHandle resultOk;
    private final MethodHandle resultErrCode;
    private final MethodHandle resultErrMsg;
    private final MethodHandle resultFree;
    private final MethodHandle stringFree;

    /** Per-context upcall state, kept alive for the context's lifetime. */
    private final ConcurrentHashMap<Long, CtxState> contexts = new ConcurrentHashMap<>();

    private record CtxState(Arena arena, MemorySegment stub, Consumer<String> sink) {
    }

    public FfmNativeBridge() {
        NativeLibraryLoader.ensureLoaded();
        SymbolLookup lookup = SymbolLookup.loaderLookup();
        var ADDR = ValueLayout.ADDRESS;
        var INT = ValueLayout.JAVA_INT;

        this.version = down(lookup, "cavs_sdk_version", FunctionDescriptor.of(ADDR));
        this.abiVersion = down(lookup, "cavs_sdk_abi_version", FunctionDescriptor.of(ADDR));
        this.capabilities = down(lookup, "cavs_sdk_capabilities_json", FunctionDescriptor.of(ADDR));
        this.ctxNew = down(lookup, "cavs_context_new", FunctionDescriptor.of(ADDR, ADDR));
        this.ctxFree = down(lookup, "cavs_context_free", FunctionDescriptor.ofVoid(ADDR));
        this.setProgress = down(lookup, "cavs_context_set_progress_callback",
                FunctionDescriptor.of(INT, ADDR, ADDR, ADDR));
        this.execute = down(lookup, "cavs_execute_json", FunctionDescriptor.of(ADDR, ADDR, ADDR, ADDR));
        this.startJson = down(lookup, "cavs_start_json", FunctionDescriptor.of(ADDR, ADDR, ADDR, ADDR));
        this.jobPoll = down(lookup, "cavs_job_poll", FunctionDescriptor.of(ADDR, ADDR));
        this.jobCancel = down(lookup, "cavs_job_cancel", FunctionDescriptor.of(INT, ADDR));
        this.jobFree = down(lookup, "cavs_job_free", FunctionDescriptor.ofVoid(ADDR));
        this.resultJson = down(lookup, "cavs_result_json", FunctionDescriptor.of(ADDR, ADDR));
        this.resultOk = down(lookup, "cavs_result_ok", FunctionDescriptor.of(INT, ADDR));
        this.resultErrCode = down(lookup, "cavs_result_error_code", FunctionDescriptor.of(ADDR, ADDR));
        this.resultErrMsg = down(lookup, "cavs_result_error_message", FunctionDescriptor.of(ADDR, ADDR));
        this.resultFree = down(lookup, "cavs_result_free", FunctionDescriptor.ofVoid(ADDR));
        this.stringFree = down(lookup, "cavs_string_free", FunctionDescriptor.ofVoid(ADDR));
    }

    private static MethodHandle down(SymbolLookup lookup, String name, FunctionDescriptor fd) {
        MemorySegment sym = lookup.find(name)
                .orElseThrow(() -> new UnsatisfiedLinkError("cavs: missing native symbol " + name));
        return LINKER.downcallHandle(sym, fd);
    }

    @Override
    public String version() {
        return staticString(version);
    }

    @Override
    public String abiVersion() {
        return staticString(abiVersion);
    }

    @Override
    public String capabilitiesJson() {
        try {
            MemorySegment ptr = (MemorySegment) capabilities.invoke();
            String s = readString(ptr);
            stringFree.invoke(ptr);
            return s;
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    @Override
    public long contextNew() {
        try {
            MemorySegment ctx = (MemorySegment) ctxNew.invoke(MemorySegment.NULL);
            return ctx.address();
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    @Override
    public void contextFree(long ctx) {
        CtxState st = contexts.remove(ctx);
        try {
            ctxFree.invoke(MemorySegment.ofAddress(ctx));
        } catch (Throwable t) {
            throw sneaky(t);
        } finally {
            if (st != null) {
                st.arena().close();
            }
        }
    }

    @Override
    public void setProgressCallback(long ctx, Consumer<String> sink) {
        try {
            CtxState prev = contexts.remove(ctx);
            if (prev != null) {
                prev.arena().close();
            }
            if (sink == null) {
                setProgress.invoke(MemorySegment.ofAddress(ctx), MemorySegment.NULL, MemorySegment.NULL);
                return;
            }
            Arena arena = Arena.ofShared();
            // void trampoline(const char* eventJson, void* userData)
            MethodHandle handler = MethodHandles.lookup()
                    .bind(new Trampoline(sink), "onEvent",
                            MethodType.methodType(void.class, MemorySegment.class, MemorySegment.class));
            MemorySegment stub = LINKER.upcallStub(
                    handler, FunctionDescriptor.ofVoid(ValueLayout.ADDRESS, ValueLayout.ADDRESS), arena);
            contexts.put(ctx, new CtxState(arena, stub, sink));
            setProgress.invoke(MemorySegment.ofAddress(ctx), stub, MemorySegment.NULL);
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    @Override
    public String executeSync(long ctx, String operation, String requestJson) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment op = arena.allocateFrom(operation);
            MemorySegment req = arena.allocateFrom(requestJson);
            MemorySegment result = (MemorySegment) execute.invoke(MemorySegment.ofAddress(ctx), op, req);
            return drainResult(result);
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    @Override
    public long startJob(long ctx, String operation, String requestJson) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment op = arena.allocateFrom(operation);
            MemorySegment req = arena.allocateFrom(requestJson);
            MemorySegment job = (MemorySegment) startJson.invoke(MemorySegment.ofAddress(ctx), op, req);
            return job.address();
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    @Override
    public String pollJob(long job) {
        try {
            MemorySegment result = (MemorySegment) jobPoll.invoke(MemorySegment.ofAddress(job));
            if (result.address() == 0) {
                return null;
            }
            return drainResult(result);
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    @Override
    public void cancelJob(long job) {
        try {
            jobCancel.invoke(MemorySegment.ofAddress(job));
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    @Override
    public void freeJob(long job) {
        try {
            jobFree.invoke(MemorySegment.ofAddress(job));
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    /** Read the result envelope JSON and free the native result. */
    private String drainResult(MemorySegment result) throws Throwable {
        if (result.address() == 0) {
            return "{\"ok\":false,\"error\":{\"code\":\"CAVS-E-INTERNAL\","
                    + "\"message\":\"native returned null result\",\"recoverable\":false}}";
        }
        MemorySegment json = (MemorySegment) resultJson.invoke(result);
        String envelope = readString(json);
        resultFree.invoke(result);
        return envelope;
    }

    private String staticString(MethodHandle h) {
        try {
            return readString((MemorySegment) h.invoke());
        } catch (Throwable t) {
            throw sneaky(t);
        }
    }

    /** Read a NUL-terminated UTF-8 C string from a returned pointer. */
    private static String readString(MemorySegment ptr) {
        if (ptr == null || ptr.address() == 0) {
            return null;
        }
        return ptr.reinterpret(Long.MAX_VALUE).getString(0);
    }

    private static RuntimeException sneaky(Throwable t) {
        if (t instanceof RuntimeException re) {
            return re;
        }
        return new RuntimeException(t);
    }

    /** Holds a Java sink and adapts the native (char*, void*) callback to it. */
    private static final class Trampoline {
        private final Consumer<String> sink;

        Trampoline(Consumer<String> sink) {
            this.sink = sink;
        }

        @SuppressWarnings("unused") // bound via MethodHandle
        void onEvent(MemorySegment eventJson, MemorySegment userData) {
            String s = readString(eventJson);
            if (s != null) {
                sink.accept(s);
            }
        }
    }
}

File: sdks/kotlin/src/main/kotlin/com/cavs/sdk/nativebridge/FfmNativeBridge.kt
package com.cavs.sdk.nativebridge

import java.lang.foreign.Arena
import java.lang.foreign.FunctionDescriptor
import java.lang.foreign.Linker
import java.lang.foreign.MemorySegment
import java.lang.foreign.SymbolLookup
import java.lang.foreign.ValueLayout
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType
import java.util.concurrent.ConcurrentHashMap

/**
 * [NativeBridge] backed by the Java Foreign Function & Memory API (JEP 454,
 * finalized in Java 22). It needs no third-party dependency: the whole
 * binding is JDK-native.
 *
 * Downcalls use [MethodHandle.invokeWithArguments] rather than the
 * signature-polymorphic `invoke`/`invokeExact` (which Kotlin cannot emit);
 * for CAVS's coarse, file-system-heavy operations the reflective dispatch
 * overhead is negligible.
 */
class FfmNativeBridge : NativeBridge {

    // Must run before the downcall-handle initializers below resolve symbols.
    init {
        NativeLibraryLoader.ensureLoaded()
    }

    private val version = down("cavs_sdk_version", FunctionDescriptor.of(ADDR))
    private val abiVersion = down("cavs_sdk_abi_version", FunctionDescriptor.of(ADDR))
    private val capabilities = down("cavs_sdk_capabilities_json", FunctionDescriptor.of(ADDR))
    private val ctxNew = down("cavs_context_new", FunctionDescriptor.of(ADDR, ADDR))
    private val ctxFree = down("cavs_context_free", FunctionDescriptor.ofVoid(ADDR))
    private val setProgress =
        down("cavs_context_set_progress_callback", FunctionDescriptor.of(INT, ADDR, ADDR, ADDR))
    private val execute = down("cavs_execute_json", FunctionDescriptor.of(ADDR, ADDR, ADDR, ADDR))
    private val startJson = down("cavs_start_json", FunctionDescriptor.of(ADDR, ADDR, ADDR, ADDR))
    private val jobPoll = down("cavs_job_poll", FunctionDescriptor.of(ADDR, ADDR))
    private val jobCancel = down("cavs_job_cancel", FunctionDescriptor.of(INT, ADDR))
    private val jobFree = down("cavs_job_free", FunctionDescriptor.ofVoid(ADDR))
    private val resultJson = down("cavs_result_json", FunctionDescriptor.of(ADDR, ADDR))
    private val resultFree = down("cavs_result_free", FunctionDescriptor.ofVoid(ADDR))
    private val stringFree = down("cavs_string_free", FunctionDescriptor.ofVoid(ADDR))

    /** Per-context upcall state, kept alive for the context's lifetime. */
    private class CtxState(val arena: Arena, @Suppress("unused") val stub: MemorySegment)

    private val contexts = ConcurrentHashMap<Long, CtxState>()

    override fun version(): String = readString(version.call() as MemorySegment)!!

    override fun abiVersion(): String = readString(abiVersion.call() as MemorySegment)!!

    override fun capabilitiesJson(): String {
        val ptr = capabilities.call() as MemorySegment
        val s = readString(ptr) ?: ""
        stringFree.call(ptr)
        return s
    }

    override fun contextNew(): Long = (ctxNew.call(MemorySegment.NULL) as MemorySegment).address()

    override fun contextFree(ctx: Long) {
        val state = contexts.remove(ctx)
        try {
            ctxFree.call(MemorySegment.ofAddress(ctx))
        } finally {
            state?.arena?.close()
        }
    }

    override fun setProgressCallback(ctx: Long, sink: ((String) -> Unit)?) {
        contexts.remove(ctx)?.arena?.close()
        if (sink == null) {
            setProgress.call(MemorySegment.ofAddress(ctx), MemorySegment.NULL, MemorySegment.NULL)
            return
        }
        val arena = Arena.ofShared()
        val handler = MethodHandles.lookup().bind(
            Trampoline(sink),
            "onEvent",
            MethodType.methodType(Void.TYPE, MemorySegment::class.java, MemorySegment::class.java),
        )
        val stub = LINKER.upcallStub(
            handler,
            FunctionDescriptor.ofVoid(ValueLayout.ADDRESS, ValueLayout.ADDRESS),
            arena,
        )
        contexts[ctx] = CtxState(arena, stub)
        setProgress.call(MemorySegment.ofAddress(ctx), stub, MemorySegment.NULL)
    }

    override fun executeSync(ctx: Long, operation: String, requestJson: String): String =
        Arena.ofConfined().use { arena ->
            val op = arena.allocateFrom(operation)
            val req = arena.allocateFrom(requestJson)
            drainResult(execute.call(MemorySegment.ofAddress(ctx), op, req) as MemorySegment)
        }

    override fun startJob(ctx: Long, operation: String, requestJson: String): Long =
        Arena.ofConfined().use { arena ->
            val op = arena.allocateFrom(operation)
            val req = arena.allocateFrom(requestJson)
            (startJson.call(MemorySegment.ofAddress(ctx), op, req) as MemorySegment).address()
        }

    override fun pollJob(job: Long): String? {
        val result = jobPoll.call(MemorySegment.ofAddress(job)) as MemorySegment
        return if (result.address() == 0L) null else drainResult(result)
    }

    override fun cancelJob(job: Long) {
        jobCancel.call(MemorySegment.ofAddress(job))
    }

    override fun freeJob(job: Long) {
        jobFree.call(MemorySegment.ofAddress(job))
    }

    /** Read the result envelope JSON and free the native result. */
    private fun drainResult(result: MemorySegment): String {
        if (result.address() == 0L) {
            return """{"ok":false,"error":{"code":"CAVS-E-INTERNAL",""" +
                """"message":"native returned null result","recoverable":false}}"""
        }
        val envelope = readString(resultJson.call(result) as MemorySegment) ?: ""
        resultFree.call(result)
        return envelope
    }

    /** Holds a Kotlin sink and adapts the native (char*, void*) callback. */
    private class Trampoline(private val sink: (String) -> Unit) {
        @Suppress("unused") // bound via MethodHandle
        fun onEvent(eventJson: MemorySegment, @Suppress("UNUSED_PARAMETER") userData: MemorySegment) {
            readString(eventJson)?.let(sink)
        }
    }

    private companion object {
        val LINKER: Linker = Linker.nativeLinker()
        val ADDR: ValueLayout = ValueLayout.ADDRESS
        val INT: ValueLayout = ValueLayout.JAVA_INT

        fun down(name: String, fd: FunctionDescriptor): MethodHandle {
            val symbol = SymbolLookup.loaderLookup().find(name)
                .orElseThrow { UnsatisfiedLinkError("cavs: missing native symbol $name") }
            return LINKER.downcallHandle(symbol, fd)
        }

        /** Invoke a downcall handle without signature polymorphism. */
        fun MethodHandle.call(vararg args: Any?): Any? =
            try {
                invokeWithArguments(*args)
            } catch (t: Throwable) {
                throw (t as? RuntimeException ?: RuntimeException(t))
            }

        /** Read a NUL-terminated UTF-8 C string from a returned pointer. */
        fun readString(ptr: MemorySegment?): String? =
            if (ptr == null || ptr.address() == 0L) null
            else ptr.reinterpret(Long.MAX_VALUE).getString(0)
    }
}
